src/zap/zapinfo.cpp

```diff
@@ -2136,19 +2136,55 @@ DWORD FilterMethodAttribsForIsSupported(DWORD attribs, CORINFO_METHOD_HANDLE ftn
                     )
                 )
             )
+        {
             return attribs;
-#endif // defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+        }
 
-        // If it's another get_IsSupported method on some other ISA class, do not report as intrinsic
-        if (isX86intrinsic
-            || strcmp(namespaceName, "System.Runtime.Intrinsics.Arm.Arm64") == 0
-            || strcmp(namespaceName, "System.Runtime.Intrinsics") == 0)
+        // If it's an intrinsic that requires VEX encoding, do not report as intrinsic
+        // to force this to become a regular method call.
+        // We don't allow RyuJIT to use VEX encoding at AOT compilation time, so these
+        // cannot be pregenerated. Not reporting them as intrinsic will make sure
+        // it will do the right thing at runtime (the called method will be JITted).
+        // It will be slower, but correct.
+        if (fIsX86intrinsic
+            && (
+                strcmp(className, "Avx") == 0 || strcmp(className, "Fma") == 0 || strcmp(className, "Avx2") == 0 || strcmp(className, "Bmi1") == 0 || strcmp(className, "Bmi2") == 0
+                || (
+                    strcmp(className, "X64") == 0
+                    && (
+                        strcmp(enclosingClassName, "Bmi1") == 0 || strcmp(enclosingClassName, "Bmi2") == 0
+                        )
+                    )
+                )
+            )
         {
-            attribs &= ~CORINFO_FLG_JIT_INTRINSIC;
+            // We do want the IsSupported for VEX instructions to be recognized as intrinsic so that the
+            // potentially worse quality code doesn't actually run until tiered JIT starts
+            // kicking in and recompiling methods. Reporting this as intrinsic makes RyuJIT expand it
+            // into `return false`.
+            if (fIsGetIsSupportedMethod)
+                return attribs;
+
+            // Treat as a regular method call (into a JITted method).
+            return (attribs & ~CORINFO_FLG_JIT_INTRINSIC) | CORINFO_FLG_DONT_INLINE;
```

Comment:
Is this going to be correct? The calls are recursive and the `mustExpand` support is dependent on `CORINFO_FLG_JIT_INTRINSIC` being set.

So, if it is recursive and `CORINFO_FLG_JIT_INTRINSIC` isn't set, it will be AOT'd as if it was a regular recursive call...

Reply:
What we want is the caller to treat these as regular, non-inline method calls and for the JIT to not AOT the method bodies of the intrinsics themselves.

Reply:
Isn't that what `CORINFO_FLG_DONT_INLINE` will fix up?

The recursive call within the actual intrinsic method body will never go through here because we don't AOT compile them.

Reply:
We have a compiler check for "is compiling for R2R/crossgen/AOT", right?

It might be simpler to just update the logic in `Compiler::impHWIntrinsic` method (see https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsicxarch.cpp#L61-L70) to do the following
```diff
    bool isIsaSupported = comp->compSupports(isa) && comp->compSupportsHWIntrinsic(isa);

    if (strcmp(methodName, "get_IsSupported") == 0)
    {
        return isIsaSupported ? NI_IsSupported_True : NI_IsSupported_False;
    }
    else if (!isIsaSupported)
    {
+       if (isCompilingForAOT)
+       {
+           return nullptr;
+       }
        return NI_Throw_PlatformNotSupportedException;
    }
```

Reply:
I believe that would allow us to treat the AVX/AVX2/FMA/BMI1/BMI2 code paths as normal (you don't need to mark them as `DONT_INLINE` or strip the `JIT_INTRINSIC` flag).

The existing logic for not expanding method bodies would kick-in here: https://github.com/dotnet/coreclr/pull/24917/commits/6b5b161733fb19ee92853b724cb10a1de395b12c#diff-04a590b55693db432b3a24d0d9a59cceR451

And the added check would ensure that not supported ISAs (which these won't be due to us not setting the compiler flags in the zapper here: https://github.com/dotnet/coreclr/pull/24917/commits/6b5b161733fb19ee92853b724cb10a1de395b12c#diff-efefdb418a1b8be4f78c7c9aedf91cb4L1212) will return `nullptr` (forcing them to be emitted as regular method calls).

Reply:
> Isn't that what CORINFO_FLG_DONT_INLINE will fix up?

I don't believe so. Intrinsics are special and are handled before the inlining logic.

Reply:
@tannergooding is correct here - the intrinsics are recognized during importing, prior to the call being considered for inlining.

I think it makes some sense to unconditionally fail to expand (i.e. `return nullptr`) for these when pre-jitting (aka AOT compiling). I don't think the JIT can distinguish when it's compiling SPC.dll vs. something else, but I don't think it makes a difference in practice, since any intrinsics that are called under the check for `IsSupported` will fail to pre-jit anyway.

Reply:
> I don't believe so. Intrinsics are special and are handled before the inlining logic.

But this line strips the intrinsic bit from the method. It should be treated as a simple method call, no?

That's what we want for BMI and friends - keep IsSupported untouched and reported as an intrinsic so that it expands to `false`, but treat everything else as a regular non-inlined method call - the fact that the IL is recursive shouldn't matter because RyuJIT is not going to look at it due to the no inline flag, no?

Reply:
> the fact that the IL is recursive shouldn't matter because RyuJIT is not going to look at it due to the no inline flag, no?

The problem ends up coming in when we go to AOT those methods. You stripped the `JIT_INTRINSIC` flag so the logic for "skip generating hardware intrinsic method bodies" won't kick in. Which means we will AOT the BMI and friends method bodies as regular recursive calls.

Reply:
> The problem ends up coming in when we go to AOT those methods.

Attributes in that spot are retrieved from the standard JitInterface - we don't call the `getMethodAttribs` that messes with the intrinsic bit there.

Precompiling method bodies on e.g. the Bmi1 class would be wrong.

Reply:
Ah, ok. That definitely wasn't clear to me. A comment calling it out might be helpful.